src/sync.rs

```diff
@@ -259,38 +307,49 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         reader: ThreadId,
     ) {
         let this = self.eval_context_mut();
-        assert!(this.rwlock_is_write_locked(id), "queueing on not write locked lock");
+        assert!(this.rwlock_is_write_locked(id), "read-queueing on not write locked rwlock");
         this.machine.threads.sync.rwlocks[id].reader_queue.push_back(reader);
         this.block_thread(reader);
     }
 
-    #[inline]
-    /// Take a reader out the queue waiting for the lock.
-    /// Returns `true` if some thread got the rwlock.
-    fn rwlock_dequeue_and_lock_reader(&mut self, id: RwLockId) -> bool {
-        let this = self.eval_context_mut();
-        if let Some(reader) = this.machine.threads.sync.rwlocks[id].reader_queue.pop_front() {
-            this.unblock_thread(reader);
-            this.rwlock_reader_lock(id, reader);
-            true
-        } else {
-            false
-        }
-    }
-
     #[inline]
     /// Lock by setting the writer that owns the lock.
     fn rwlock_writer_lock(&mut self, id: RwLockId, writer: ThreadId) {
         let this = self.eval_context_mut();
         assert!(!this.rwlock_is_locked(id), "the rwlock is already locked");
+        trace!("rwlock_writer_lock: {:?} now held by {:?}", id, writer);
         this.machine.threads.sync.rwlocks[id].writer = Some(writer);
     }
 
     #[inline]
     /// Try to unlock by removing the writer.
-    fn rwlock_writer_unlock(&mut self, id: RwLockId) -> Option<ThreadId> {
+    fn rwlock_writer_unlock(&mut self, id: RwLockId, expected_writer: ThreadId) -> bool {
         let this = self.eval_context_mut();
-        this.machine.threads.sync.rwlocks[id].writer.take()
+        let rwlock = &mut this.machine.threads.sync.rwlocks[id];
+        if let Some(current_writer) = rwlock.writer {
+            if current_writer != expected_writer {
+                // Only the owner can unlock the rwlock.
+                return false;
```

Comment:
Shouldn't this be an error?

Reply:
We leave error reporting to the caller. Mutexes do the same.